contrib/google.golang.org/api/internal/gen_endpoints/main.go

```diff
@@ -0,0 +1,270 @@
+				}
+				es = append(es, res...)
+			}
+		}
+
+		return nil
+	})
+	checkError(err)
+
+	f, err := os.Create(outFile)
+	checkError(err)
+	defer f.Close()
+
+	sort.Slice(es, func(i, j int) bool {
+		return es[i].String() < es[j].String()
+	})
+
+	log.Printf("Generating templated Go code in %s...\n", outFile)
+	template.Must(template.New("").Parse(tpl)).Execute(f, map[string]interface{}{
+		"Endpoints": es,
+		"Year":      time.Now().Year(),
+	})
+	log.Println("Done!")
+}
+
+func checkError(err error) {
+	if err != nil {
+		log.Fatalln(err)
+	}
+}
+
+func handleResource(def *APIDefinition, resource *APIResource) ([]Endpoint, error) {
+	var es []Endpoint
+	if resource.Methods != nil {
+		for _, method := range resource.Methods {
+			mes, err := handleMethod(def, method)
+			if err != nil {
+				return nil, err
+			}
+			es = append(es, mes...)
+		}
+	}
+	if resource.Resources != nil {
+		for _, child := range resource.Resources {
+			res, err := handleResource(def, child)
+			if err != nil {
+				return nil, err
+			}
+			es = append(es, res...)
+		}
+	}
+	return es, nil
+}
+
+func handleMethod(def *APIDefinition, method *APIMethod) ([]Endpoint, error) {
+	u, err := url.Parse(def.RootURL)
+	if err != nil {
+		return nil, err
+	}
+	hostname := u.Hostname()
+
+	path := method.FlatPath
+	if path == "" {
+		path = method.Path
+	}
+	path = def.BasePath + path
+
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	uritpl, err := uritemplate.New(path)
+	if err != nil {
+		return nil, err
+	}
+	return []Endpoint{{
+		Hostname:     hostname,
+		HTTPMethod:   method.HTTPMethod,
+		PathTemplate: path,
+		PathMatcher:  uritpl.Regexp(),
+		ServiceName:  "google." + def.Name,
+		ResourceName: method.ID,
+	}}, nil
+}
+
+func downloadGoogleApiSrc() string {
+	zipUrl := downloadUrl()
+	dir := os.TempDir()
+
+	zipFile := path.Join(dir, "google-api-go-client.zip")
+	defer os.Remove(zipFile)
+	dst := path.Join(dir, "google-api-go-client")
+	err := os.Mkdir(dst, os.ModePerm)
+	checkError(err)
+
+	log.Printf("Downloading %s into %s...\n", zipUrl, dst)
+
+	out, err := os.Create(zipFile)
+	defer out.Close()
+	resp, err := http.Get(zipUrl)
+	checkError(err)
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	checkError(err)
+
+	log.Printf("Extracting %s into %s...\n", zipFile, dst)
+
+	zf, err := zip.OpenReader(zipFile)
+	checkError(err)
+	defer zf.Close()
+
+	for _, f := range zf.File {
+		filePath := filepath.Join(dst, f.Name)
+		if !strings.HasPrefix(filePath, filepath.Clean(dst)+string(os.PathSeparator)) {
+			panic("invalid file path")
+		}
+		if f.FileInfo().IsDir() {
+			os.MkdirAll(filePath, os.ModePerm)
+			continue
+		}
+		err = os.MkdirAll(filepath.Dir(filePath), os.ModePerm)
+		checkError(err)
+
+		dstFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+		checkError(err)
+
+		fileInArchive, err := f.Open()
+		checkError(err)
+
+		_, err = io.Copy(dstFile, fileInArchive)
+		checkError(err)
+
+		dstFile.Close()
```

Comment:
should these closes be deferred before the error check?

Reply:
extracted this code into a separate function so linter does not complain about `defer` inside a loop 👍